### PR TITLE
Fix batchnorm problem with sparse matrices when fix_gamma=True

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
As title.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fix fallback logic for batchnorm operator
- [x] Block usage of fix_gamma=True when input is sparse
- [x] Unit tests

## Comments ##
Resurrection of #11631.

### CODE REVIEW

**Comment by haojin2** [comment]
@marcoabreu Seems like there's no build triggered for this: http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/activity/?branch=PR-11656. Can you take a look?

**Comment by anirudh2290** [comment]
Link to the issue: https://github.com/apache/incubator-mxnet/issues/11655

**Comment by haojin2** [comment]
@anirudh2290 That's a duplicate issue of #11654.

File: src/operator/batch_norm_v1.cc [hunk 1]
```diff
@@ -89,6 +89,9 @@ the output. It is often used during inference.
 Both ``gamma`` and ``beta`` are learnable parameters. But if ``fix_gamma`` is true,
 then set ``gamma`` to 1 and its gradient to 0.
 
+There's no sparse support for this operator, and will exhibit problematic behavior if used with
```
Comment (KellenSunderland) [inline src/operator/batch_norm_v1.cc]: Nit: "There's no sparse support for this operator, and will" -> "There's no sparse support for this operator. It will"

File: tests/python/mkl/test_mkldnn.py [hunk 2]
```diff
@@ -233,7 +233,7 @@ def check_batchnorm_training(stype):
                            mx.nd.array(beta).tostype(stype)]
             mean_std = [mx.nd.array(rolling_mean).tostype(stype), mx.nd.array(rolling_std).tostype(stype)]
 
-            test = mx.symbol.BatchNorm(data, fix_gamma=True)
+            test = mx.symbol.BatchNorm(data, fix_gamma=False)
```
Comment (marcoabreu) [inline tests/python/mkl/test_mkldnn.py]: Could you elaborate why you are modifying a test case?

Comment (haojin2) [reply tests/python/mkl/test_mkldnn.py]: I thought we had the conversation in Lai's PR yesterday, and I've created the related issue #11647

Comment (haojin2) [reply tests/python/mkl/test_mkldnn.py]: The purpose of this mkl test is to test that we can fallback to sparse matrices correctly when using MKLDNN under the legal cases for sparse matrices. The logics of exclusion of illegal cases for sparse matrices(fix_gamma=True), which is shared by both USE_MKLDNN=1 and USE_MKLDNN=0 situtaions, were already tested by the test_batchnorm_fallback test.

Comment (marcoabreu) [reply tests/python/mkl/test_mkldnn.py]: Ah I see, sorry, got a lot of PRs swirling around my head. Let me rephrase to ensure I got it right: This means that the test case with gamma=True is basically invalid until that issue has been resolved and that's why you are changing it, correct?
If yes, we're good to go. Sorry for the caused inconvenience.

Comment (haojin2) [reply tests/python/mkl/test_mkldnn.py]: Yes you got it, that's why we're only testing the necessary legal case here. Please merge this once you feel good to do so, thanks!

Comment (marcoabreu) [reply tests/python/mkl/test_mkldnn.py]: I'd like to request the review from another committer to assess the backend changes if you don't mind.

Comment (haojin2) [reply tests/python/mkl/test_mkldnn.py]: No problem. @zheng-da @eric-haibin-lin

Comment (haojin2) [reply tests/python/mkl/test_mkldnn.py]: BTW I think @anirudh2290 already approved the previous #11631

File: src/operator/nn/batch_norm.cc [hunk 3]
```diff
@@ -444,27 +445,39 @@ void BatchNormGradComputeExCPU(const nnvm::NodeAttrs &attrs,
...
 
-static inline bool backward_BatchNormStorageType(const nnvm::NodeAttrs &attrs,
-                                                 const int dev_mask,
-                                                 DispatchMode *dispatch_mode,
-                                                 std::vector<int> *in_attrs,
-                                                 std::vector<int> *out_attrs) {
-  return MKLDNNStorageType(attrs, dev_mask, true, dispatch_mode,
-                           in_attrs, out_attrs);
-}
+  if ((common::ContainsStorageType(*in_attrs, kRowSparseStorage) ||
```
Comment (eric-haibin-lin) [inline src/operator/nn/batch_norm.cc]: Change it to `!common::ContainsOnlyStorage(kDefault)`, in case we add more stype in the future.

Comment (haojin2) [reply src/operator/nn/batch_norm.cc]: Done.
